### PR TITLE
Spanner struct parameters

### PR DESCRIPTION
**Pull request by jskeet**
Definitely review one commit at a time.

Lots of room for changes if required. I'm most interested in the API of SpannerStruct.
(We may want to consider an overload to SpannerParameterCollection to make it easier to add structs. Not sure.)

### CODE REVIEW

**Comment by jskeet** [comment]
Will have a look at the CI failures. The plan will be to merge this into a new branch rather than master for now.

**Comment by jskeet** [comment]
(Removing you from assignment until I've fixed the bug, Chris. Feel free to review if you'd otherwise be twiddling your thumbs.)

**Comment by jskeet** [comment]
Okay, I now understand the issue, and part of it is easily fixed - I'll push an extra commit in a minute.
The other part is *not* easily fixed. SpannerStruct doesn't implement IEquatable, and it's hard to make it do so in a way which is useful for tests - because the value in a field can itself be a list or an array. That means our equality assertion fails in the deserialization tests.

I think the way to go is to effectively extract the deserialization tests involving structs, and hand-write the equivalents.

**Comment by jskeet** [comment]
Okay, now actually ready for review - still one commit at a time. Do ping me if anything doesn't make sense.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs [hunk 1]
```diff
@@ -333,12 +307,35 @@ public override int GetHashCode()
...
+            internal string Name { get; }
+            internal SpannerDbType Type { get; }
+
+            public StructField(string name, SpannerDbType type)
+            {
+                Name = GaxPreconditions.CheckNotNull(name, nameof(name));
+                Type = GaxPreconditions.CheckNotNull(type, nameof(type));
+            }
+
+            public override bool Equals(object obj) => obj is StructField sf ? Equals(sf) : false;
```
Comment (evildour) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs]: What about `obj is StructField sf && Equals(sf);`?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/StructParameterTests.cs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+                        Assert.True(await reader.ReadAsync());
+                        Assert.True(reader.IsDBNull(0));
+                        Assert.False(await reader.ReadAsync());
+                    }
+                }
+            }
+        }
+
+        [Fact]
+        public async Task ServerStructEquality()
```
Comment (evildour) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/StructParameterTests.cs]: Maybe add an associated equality check using the wrong values to verify the values are actually being compared?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs]: Done. (Not sure why I didn't think of that...)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerStruct.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        /// <summary>
+        /// Adds a new field to the struct with the given name, type and value.
+        /// </summary>
+        /// <param name="name">The name of the field. May be null, and does not need to be unique within the struct.
+        /// If the value is null, the <see cref="Field"/> value added will have an empty string for the name.</param>
+        /// <param name="type">The type of the field. Must not be null.</param>
+        /// <param name="value">The value of the field. May be null.</param>
+        public void Add(string name, SpannerDbType type, object value) =>
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerStruct.cs]: Done. I'll make Add(Field) return the field as well, just for consistency.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/StructParameterTests.cs (hunk 2 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/StructParameterTests.cs]: Sure. (This was basically a list of suggested tests - happy to add another one though.)

**Comment by jskeet** [comment]
@evildour: When you looked through this, did you do so sufficiently thoroughly that it's not worth @chrisdunelm doing so as well, or was it a skim?

**Comment by evildour** [comment]
I gave it a thorough review. That said, it always helps to have another set of eyes on it, but it isn't necessary.

**Comment by jskeet** [comment]
Fab - given that it's only going into a new branch anyway (not straight to master) I'll merge it, and Chris can look later with no risk of problems. Thanks!

**Pull request merged**